gateway/src/main/java/org/georchestra/gateway/filter/global/AccessLogFilter.java

```diff
@@ -0,0 +1,126 @@
+ * the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * geOrchestra is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * geOrchestra.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.georchestra.gateway.filter.global;
+
+import static org.springframework.cloud.gateway.support.ServerWebExchangeUtils.GATEWAY_REQUEST_URL_ATTR;
+import static org.springframework.cloud.gateway.support.ServerWebExchangeUtils.GATEWAY_ROUTE_ATTR;
+
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.security.Principal;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.georchestra.gateway.model.GeorchestraUsers;
+import org.georchestra.security.model.GeorchestraUser;
+import org.slf4j.MDC;
+import org.springframework.cloud.gateway.filter.GatewayFilterChain;
+import org.springframework.cloud.gateway.filter.GlobalFilter;
+import org.springframework.cloud.gateway.route.Route;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.security.authentication.AnonymousAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.web.server.ServerWebExchange;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import reactor.core.publisher.Mono;
+
+@RequiredArgsConstructor
+@Slf4j(topic = "org.georchestra.gateway.accesslog")
+public class AccessLogFilter implements GlobalFilter {
+
+    private final @NonNull AccessLogFilterConfig config;
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (config.matches(exchange.getRequest().getURI())) {
+            exchange.getResponse().beforeCommit(() -> {
+                return log(exchange);
+            });
+        }
+
+        return chain.filter(exchange);
+    }
+
+    private static final AnonymousAuthenticationToken ANNON = new AnonymousAuthenticationToken("anonymous", "anonymous",
+            List.of(new SimpleGrantedAuthority("ROLE_ANONYMOUS")));
+
+    /**
+     * @param exchange
+     */
+    private Mono<Void> log(ServerWebExchange exchange) {
+        if (!log.isInfoEnabled())
+            return Mono.empty();
+
+        return exchange.getPrincipal().switchIfEmpty(Mono.just(ANNON)).doOnNext(p -> {
+            doLog(p, exchange);
```

Comment:
Apparently, we're having 2 issues there: 
- in my experiments, I'm always getting a null Principal (then replaced by ANNON)
- but also, the flow never gets to `doLog` (something  seems to be wrong with doOnNext)

Reply:
I did several different tries with no luck:
* removing the switchIfEmpty
* replacing the doOnNext by a flatMap (as advised on one of a SO answer to a similar issue)
* removing the then() call

My last try looked like this:

```
    /**
     * @param exchange
     */
    private Mono<Void> log(ServerWebExchange exchange, GatewayFilterChain chain) {
        if (!log.isInfoEnabled())
            return Mono.empty();

        return exchange.getPrincipal().flatMap(p -> {
            doLog(p, exchange);
            return chain.filter(exchange);
        });
    }
```

But still the doLog() call is never reached.

Reply:
AFAIK, removing the switchIfEmpty would not work, since I'm under the impression that the Principal doesn't get populated and doOnNext would not react to an empty Mono.
But I admit, this all still sounds very much like chinese to me
I admit I find it a bit bothering to rely on some code that is _that_ difficult to even understand (let alone debug). Or maybe I'm just getting too old for this s#!t

Reply:
> Or maybe I'm just getting too old for this s#!t

that syntaxic sugar (is that modern java ?) almost looks like perl5 with less #@$& :)

Reply:
I wonder if there wouldn't be, also, a thread context propagation issue. By tweaking a few things (making AccessLogFilter class implement `Ordered`, commenting the `if (config.matches...` block), I've been able to insert a hard-coded MDC. But it shows on `parrallel-1` thread only while the access logs are on `reactor-http-epoll-3` thread.